File: MovieManagerXamarin2/MovieManagerXamarin2/MovieManagerXamarin2/Services/MovieService.cs
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using MovieManagerXamarin2.Models;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;

namespace MovieManagerXamarin2.Services
{
    public interface IMovieService
    {
        Task<List<Movie>> GetMovies(string userId);

        Task ToggleFavorite(string userId, string imdbId, bool favorite);

        Task<T> OnlineMovieSearch<T>(string title, int page = 1);

        Task<Movie> AddMovie(Movie movie);

        Task DeleteMovie(Movie movie);

        Task<MovieDetailResponse> GetMovieDetails(string imdbId);

        Task<List<string>> GetMovieFormats();

        Task AddFromWishlist(Movie movie);
    }

    public class MovieService : BaseService, IMovieService
    {
        private readonly LoginService _loginService;
        private readonly JsonSerializerSettings _jsonSettings;

        public MovieService()
        {
            
            _loginService = new LoginService();
            _jsonSettings = new JsonSerializerSettings
            {
                ContractResolver = new CamelCasePropertyNamesContractResolver()
            };
        }

        public async Task<List<Movie>> GetMovies(string userId)
        {
            var url = $"{ApiBaseUrl}/movies/{userId}";
            var asString = await HttpClient.GetStringAsync(url);

            return JsonConvert.DeserializeObject<List<Movie>>(asString);
        }

        public async Task ToggleFavorite(string userId, string imdbid, bool favorite)
        {
            var data = new
            {
                userId,
                imdbid,
                update = new
                {
                    favorite
                }
            };

            var content = new StringContent(JsonConvert.SerializeObject(data), Encoding.UTF8, "application/json");

            await HttpClient.PutAsync($"{ApiBaseUrl}/movie", content);
        }

        public async Task<T> OnlineMovieSearch<T>(string title, int page = 1)
        {
            var url = $"{ApiBaseUrl}/movie-search/${title}/${page}";
            var asString = await HttpClient.GetStringAsync(url);

            return JsonConvert.DeserializeObject<T>(asString);
        }

        public async Task<Movie> AddMovie(Movie movie)
        {
            var user = _loginService.GetSavedCredentials();
            var data = new
            {
                title = movie.Title,
                description = string.Empty,
                userId = user?.UserId ?? "58cb3dd6692c796b68ff33ec",
                director = movie.Director,
                imdbid = movie.ImdbId,
                wishlist = movie.Wishlist,
                format = movie.Wishlist ? "" : movie.Format //Do not set format for wishlist adds.
            };

            var asString = new StringContent(JsonConvert.SerializeObject(data, _jsonSettings), 
                Encoding.UTF8, "application/json");
            var response = await HttpClient.PostAsync($"{ApiBaseUrl}/movie", asString);

            return JsonConvert.DeserializeObject<Movie>(
                await response.Content.ReadAsStringAsync());
        }

        public async Task DeleteMovie(Movie movie)
        {
            var url = $"{ApiBaseUrl}/movie/{movie.UserId}/{movie.ImdbId}";

            await HttpClient.DeleteAsync(url);
        }

        public async Task<MovieDetailResponse> GetMovieDetails(string imdbId)
        {
            var httpResponse =
                await HttpClient.GetStringAsync($"{ApiBaseUrl}/movie-details/{imdbId}");

            return JsonConvert.DeserializeObject<MovieDetailResponse>(httpResponse);
        }

        public async Task<List<string>> GetMovieFormats()
        {
            var httpResponse = await HttpClient.GetStringAsync($"{ApiBaseUrl}/movie-formats");

            return JsonConvert.DeserializeObject<List<string>>(httpResponse);
        }

        public async Task AddFromWishlist(Movie movie)
        {
            var data = new
            {
                userId = movie.UserId,
                imdbid = movie.ImdbId,
                update = new
                {
                    wishlist = false,
                    format = movie.Format
                }
            };

            var asString = new StringContent(
                JsonConvert.SerializeObject(data), Encoding.UTF8, "applicaiton/json");

            await HttpClient.PutAsync($"{ApiBaseUrl}/movie", asString);
        }
    }
}

File: app/services/movie-service.ts
import { MovieViewModel } from "~/view-models/movie-view-model";
import { Movie, MovieDetailResponse } from "~/shared/interfaces";
import { LoginService } from './login-service';
import { BaseService } from './base-service';

let movies: any = [
    {
        _id: '1',
        title: 'The Shawshank Redemption',
        description: 'Drama',
        userId: '123',
        director: 'Frank Darabont',
        __v: 0,
        onlineId: 'tt0111161'
    },
    {
        _id: '2',
        title: 'The Big Lebowski',
        description: 'Drama',
        userId: '123',
        director: 'Joel Cohen',
        onlineId: 'tt0118715'
    },
    {
        _id: '3',
        title: 'L.A. Confidential',
        description: 'Drama',
        userId: '123',
        director: 'Curtis Hanson',
        onlineId: 'tt0119488'
    },
    {
        _id: '4',
        title: 'The Royal Tennenbaums',
        description: 'Drama',
        userId: '123',
        director: 'Wes Anderson',
        onlineId: 'tt0265666'
    }
];

export class MovieService extends BaseService {
    private useHttpService: boolean = true;
    private loginService: LoginService;

    constructor() {
        super();
        this.loginService = new LoginService();
    }

    getMovies<T>(): Promise<T> {
        if (this.useHttpService) {
            let user = this.loginService.GetSavedCredentials();
            let requestParams = {
                url: `${this.apiBaseUrl}/movies/${user.userId}`,
                method: 'GET'
            }
            return this.ProcessHttpCall<T>(requestParams);
        } else {
            return new Promise<T>((resolve, reject) => {
                resolve(movies);
            });
        }
    }

    getMovie(imdbid: string): Promise<MovieDetailResponse> {
        let user = this.loginService.GetSavedCredentials();
        let requestParams = {
            url: `${this.apiBaseUrl}/movie/${user.userId}/${imdbid}`,
            method: 'GET'
        };

        return this.ProcessHttpCall<MovieDetailResponse>(requestParams);
    }

    addMovie(movie: MovieViewModel): Promise<Movie> {
        let user = this.loginService.GetSavedCredentials();
        let data: Movie = {
            _id: undefined,
            title: movie.Title,
            description: '',
            userId: user.userId,
            director: movie.Director,
            imdbid: movie.ImdbId,
            favorite: false,
            wishlist: movie.Wishlist,
            format: movie.Wishlist ? '' : movie.Format,
            poster: movie.Poster
        };

        let requestParams = {
            url: `${this.apiBaseUrl}/movie`,
            method: 'POST',
            headers: { "Content-Type": "application/json" },
            content: JSON.stringify(data)
        };

        return this.ProcessHttpCall<Movie>(requestParams);
    }

    deleteMovie(movie: MovieViewModel): Promise<any> {
        let requestParams = {
            url: `${this.apiBaseUrl}/movie/${movie.UserId}/${movie.ImdbId}`,
            method: 'DELETE'
        };

        return this.ProcessHttpCall<any>(requestParams);
    }

    getMovieDetails<T>(onlineId: string): Promise<T> {
        let requestParams = {
            url: `${this.apiBaseUrl}/movie-details/${onlineId}`,
            method: 'GET'
        };

        return this.ProcessHttpCall<T>(requestParams);
    }

    onlineMovieSearch<T>(title: string, page: number = 1): Promise<T> {
        let requestParams = {
            url: encodeURI(`${this.apiBaseUrl}/movie-search/${title}/${page}`),
            method: 'GET'
        };
        return this.ProcessHttpCall<T>(requestParams);
    }

    toggleFavorite(userId: string, imdbid: string, favorite: boolean) {
        let data = {
            userId: userId,
            imdbid: imdbid,
            update: {favorite: favorite}
        };

        let requestParams = {
            url: `${this.apiBaseUrl}/movie`,
            method: 'PUT',
            headers: { "Content-Type": "application/json" },
            content: JSON.stringify(data)
        };

        return this.ProcessHttpCall<any>(requestParams)
    }

    toggleWishlist(userId: string, imdbid: string, wishlist: boolean, format: string) {
        let data = {
            userId: userId,
            imdbid: imdbid,
            update: {wishlist: wishlist, format: format}
        };

        let requestParams = {
            url: `${this.apiBaseUrl}/movie`,
            method: 'PUT',
            headers: { "Content-Type": "application/json" },
            content: JSON.stringify(data)
        };

        return this.ProcessHttpCall<any>(requestParams);
    }

    updateFormats(userId: string, imdbid: string, formats: string) {
        const data = {
            userId: userId,
            imdbid: imdbid,
            update: {
                format: formats
            }
        };

        let requestParams = {
            url: `${this.apiBaseUrl}/movie`,
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            content: JSON.stringify(data)
        };

        return this.ProcessHttpCall<any>(requestParams);
    }

    updateRating(userId: string, imdbid: string, rating: number) {
        const data = {
            userId: userId,
            imdbid: imdbid,
            update: {
                rating: rating
            }
        };

        let requestParams = {
            url: `${this.apiBaseUrl}/movie`,
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            content: JSON.stringify(data)
        };

        return this.ProcessHttpCall<any>(requestParams);
    }

    updateMovies() {
        this.getMovies<Movie[]>().then(movies => {
            for (let movie of movies) {
                let data = {
                    userId: movie.userId,
                    imdbid: movie.imdbid,
                    update: {format: 'Blu-ray'}
                };

                let request = {
                    url: `${this.apiBaseUrl}/movie`,
                    method: 'PATCH',
                    headers: {"Content-Type": "application/json"},
                    content: JSON.stringify(data)
                };

                this.ProcessHttpCall<any>(request);
            }
        });
    }

    getFormats(): Promise<string[]> {
        let request = {
            url: `${this.apiBaseUrl}/movie-formats`,
            method: 'GET'
        }

        return this.ProcessHttpCall<string[]>(request);
    }
}